### PR TITLE
Tighten up the `gen_docs.dart` script.

### PR DESCRIPTION
**Pull request by lrhn**
Move some functionality into helper functions, to make the overall logic easier to read. Extracted constants, added comments.

Make the script work when called from any directory (finds package root relative to `Platform.script`, instead of assuming current working directory is the root.)

Remove dependency on `package:collection` which was only used for one function, in code that was restructured.

General functionality should still be the same (the generated output and JSON cache haven't changed).

### CODE REVIEW

File: tool/gen_docs.dart [hunk 1]
```diff
@@ -5,98 +5,208 @@
 import 'dart:convert';
 import 'dart:io';
 
-import 'package:collection/collection.dart';
```
Comment (devoncarew) [inline tool/gen_docs.dart:8]: 👍

**Comment by lrhn** [comment]
(I fell over the script during something else and wanted to change a little thing. Then it snowballed.)

I'm not entirely sure I understand the point of the "validate" mode, which checks that the *cached* JSON and generated Markdown are in sync, which they should be when both are written in the same runs of the tool.

It would make more sense to check that the cached JSON is up-to-date with the remote source-of-truth, because if not, we should update both local files, *or* to generate *and write* the markdown if it's not in sync with the cached JSON, which can happen if modifying the tool itself. 
Detecting the markdown being out of sync, but only being able to update it by fetching the remote JSON first, is a combination I can find no use for.

**Review by devoncarew — APPROVED** [review]
> Make the script work when called from any directory (finds package root relative to Platform.script, instead of assuming current working directory is the root.)

I tend to enforce calling from a known location (the package root, the repo root, ...) and fast fail if run from elsewhere; but no particular objections to this technique.

**Comment by devoncarew** [comment]
> I'm not entirely sure I understand the point of the "validate" mode, which checks that the _cached_ JSON and generated Markdown are in sync, which they should be when both are written in the same runs of the tool.

Yeah, the way I look at it, we want:

- the lints json metadata file to be updated periodically
- whenever the rule sets change (add a rule, remove a rule, ...), we want the markdown file to be updated, and we want to verify that on the bots so the PR author doesn't forget

Probably that means that in the 'verify only' mode, we can skip download a new json file and just use what's in the repo.

**Comment by lrhn** [comment]
I'll consider a separate change to the logic, where the "verify" mode becomes "rebuild without downloading", so it actually saves the new generated documentation. (And tells you if there are any changes.)

Landing this for now.

**Pull request merged**